File: workflows/org.linxira.benchmark-python/src/implementations/expression_pca_v1.py
"""expression.pca.v1 — independent NumPy implementation.

A line-for-line port of the Rust engine's PCA
(engine/crates/linxira-bio-core/src/expression.rs) so the two backends are
comparable field by field at the 1e-6 relative tolerance:

* the matrix reader follows the same rules (delimiter probed from the header,
  all fields trimmed, non-empty unique feature ids and sample names, missing
  values ``""``/``.``/``na``/``nan`` rejected, strictly finite floats)
* centering per feature, optional per-feature scaling by the sample standard
  deviation, constant features zeroed and counted
* the components come from the same deterministic power iteration on the
  sample Gram matrix (deterministic sinusoidal start vector, Gram-Schmidt
  against earlier components, the same convergence thresholds), so the
  eigenvector *signs* agree with the engine: the element with the largest
  absolute value is made positive
* loadings are sorted the same way (loading descending, feature ascending on
  ties) and the top-10 positive/negative lists carry that exact order

scikit-learn is deliberately not used here: its SVD sign convention differs,
and the point of this backend is a field-for-field comparison against the
engine. ``numpy.linalg`` is used only for elementwise math and ``math`` for
the sinusoidal start vector, both identical to the Rust port.
"""

from __future__ import annotations

import csv
import gzip
import math
from pathlib import Path
from typing import Any

import numpy as np

CAPABILITY = "expression.pca.v1"
INPUT_ROLES = ("matrix",)
PARAMETERS: tuple[str, ...] = ("components", "scale_features")
GZIP_MAGIC = b"\x1f\x8b"
MAX_COMPONENTS = 64
F64_EPSILON = 2.220446049250313e-16

_WARMUP_LINES = 8


class ImplementationError(ValueError):
    """An input the native engine would also reject, reported the same way."""


def software() -> list[dict[str, str]]:
    return [{"name": "NumPy", "version": np.__version__, "package_id": "numpy"}]


def _open_text(path: Path):
    with path.open("rb") as probe:
        magic = probe.read(2)
    if magic == GZIP_MAGIC:
        return gzip.open(path, "rt", encoding="utf-8", newline="")
    return path.open("r", encoding="utf-8", newline="")


def _delimiter(header_line: str) -> str:
    tabs = header_line.count("\t")
    commas = header_line.count(",")
    return "\t" if tabs > commas else ","


def _is_missing(value: str) -> bool:
    return (
        value == ""
        or value == "."
        or value.lower() in ("na", "nan")
    )


def _read_matrix(path: Path) -> tuple[list[str], list[str], list[list[float]]]:
    with _open_text(path) as handle:
        first_line = ""
        for line in handle:
            if line.strip():
                first_line = line
                break
        delimiter = _delimiter(first_line)
        handle.seek(0)
        rows_iter = csv.reader(handle, delimiter=delimiter)
        header = next(rows_iter, None)
        if header is None:
            raise ImplementationError("expression matrix is empty")
        headers = [cell.strip() for cell in header]
        if len(headers) < 2 or not headers[0]:
            raise ImplementationError(
                "expected a named feature identifier column and at least one sample"
            )
        sample_names = headers[1:]
        if any(not name for name in sample_names) or len(set(sample_names)) != len(sample_names):
            raise ImplementationError("sample names must be non-empty and unique")
        feature_ids: list[str] = []
        values: list[list[float]] = []
        seen_features: set[str] = set()
        for record_index, record in enumerate(rows_iter):
            row_number = record_index + 1
            if len(record) != len(headers):
                raise ImplementationError(
                    f"record {row_number} has {len(record)} fields, expected {len(headers)}"
                )
            feature_id = record[0].strip()
            if not feature_id or feature_id in seen_features:
                raise ImplementationError(
                    f"feature identifier {feature_id!r} is empty or duplicated"
                )
            seen_features.add(feature_id)
            row: list[float] = []
            for sample_index, cell in enumerate(record[1:]):
                value = cell.strip()
                if _is_missing(value):
                    raise ImplementationError(
                        f"sample {sample_names[sample_index]!r} contains a missing value; "
                        "impute or filter it before this analysis"
                    )
                try:
                    parsed = float(value)
                except ValueError:
                    raise ImplementationError(
                        f"sample {sample_names[sample_index]!r} contains non-numeric "
                        f"value {value!r}"
                    ) from None
                if not math.isfinite(parsed):
                    raise ImplementationError(
                        f"sample {sample_names[sample_index]!r} contains a non-finite value"
                    )
                row.append(parsed)
            feature_ids.append(feature_id)
            values.append(row)
    if not values:
        raise ImplementationError("expression matrix contains no feature rows")
    return feature_ids, sample_names, values


def _centered_rows(
    values: list[list[float]], scale: bool
) -> tuple[list[list[float]], int]:
    constant_count = 0
    rows: list[list[float]] = []
    for row in values:
        mean = sum(row) / len(row)
        centered = [value - mean for value in row]
        sum_squares = sum(value * value for value in centered)
        if sum_squares <= F64_EPSILON:
            constant_count += 1
            centered = [0.0] * len(centered)
        elif scale:
            denominator = float(max(len(row) - 1, 1))
            standard_deviation = math.sqrt(sum_squares / denominator)
            centered = [value / standard_deviation for value in centered]
        rows.append(centered)
    return rows, constant_count


def _dot(left: list[float], right: list[float]) -> float:
    return sum(a * b for a, b in zip(left, right))


def _orthogonalize(vector: list[float], basis: list[list[float]]) -> None:
    for direction in basis:
        projection = _dot(vector, direction)
        for index, direction_value in enumerate(direction):
            vector[index] -= projection * direction_value


def _normalize(vector: list[float]) -> float:
    norm = math.sqrt(_dot(vector, vector))
    if norm > F64_EPSILON:
        for index in range(len(vector)):
            vector[index] /= norm
    return norm


def _gram(rows: list[list[float]], vector: list[float], denominator: float) -> list[float]:
    result = [0.0] * len(vector)
    for row in rows:
        projection = _dot(row, vector) / denominator
        for index, row_value in enumerate(row):
            result[index] += row_value * projection
    return result


def _leading_eigenpairs(
    rows: list[list[float]], component_count: int, denominator: float
) -> list[tuple[float, list[float]]]:
    sample_count = len(rows[0])
    eigenvectors: list[list[float]] = []
    eigenpairs: list[tuple[float, list[float]]] = []
    for component in range(component_count):
        vector = [
            math.sin((index + 1) * (component + 2))
            + math.cos(((index + 1) * (component + 2)) * 0.37)
            for index in range(sample_count)
        ]
        _orthogonalize(vector, eigenvectors)
        if _normalize(vector) <= F64_EPSILON:
            found = False
            for basis in range(sample_count):
                vector = [0.0] * sample_count
                vector[basis] = 1.0
                _orthogonalize(vector, eigenvectors)
                if _normalize(vector) > F64_EPSILON:
                    found = True
                    break
            if not found:
                break
        for _ in range(500):
            nxt = _gram(rows, vector, denominator)
            _orthogonalize(nxt, eigenvectors)
            if _normalize(nxt) <= 1e-14:
                break
            alignment = abs(_dot(vector, nxt))
            vector = nxt
            if 1.0 - alignment < 1e-11:
                break
        projected = _gram(rows, vector, denominator)
        eigenvalue = _dot(vector, projected)
        if not math.isfinite(eigenvalue) or eigenvalue <= 1e-12:
            break
        # The engine makes the largest-magnitude element positive. Rust's
        # `max_by` returns the LAST maximum on ties, so scan with >= to keep
        # the tie rule identical across backends.
        largest_index = 0
        for index in range(1, len(vector)):
            if abs(vector[index]) >= abs(vector[largest_index]):
                largest_index = index
        if vector[largest_index] < 0.0:
            vector = [-value for value in vector]
        eigenvectors.append(list(vector))
        eigenpairs.append((eigenvalue, vector))
    return eigenpairs


def run(inputs: dict[str, Path], parameters: dict[str, Any]) -> dict[str, Any]:
    components_requested = int(parameters.get("components", 2))
    if components_requested <= 0:
        raise ImplementationError("components must be at least 1")
    scale_features = bool(parameters.get("scale_features", False))
    feature_ids, sample_names, values = _read_matrix(inputs["matrix"])
    if len(sample_names) < 2:
        raise ImplementationError("PCA requires at least two samples")

    rows, constant_features = _centered_rows(values, scale_features)
    denominator = float(len(sample_names) - 1)
    total_variance = (
        sum(value * value for row in rows for value in row) / denominator
    )
    if total_variance <= F64_EPSILON:
        raise ImplementationError(
            "PCA requires at least one feature with non-zero variance"
        )

    component_limit = min(
        components_requested,
        len(sample_names) - 1,
        len(feature_ids),
    )
    eigenpairs = _leading_eigenpairs(rows, component_limit, denominator)
    if not eigenpairs:
        raise ImplementationError("PCA could not resolve a non-zero component")

    samples = {sample: [] for sample in sample_names}
    components: list[dict[str, Any]] = []
    for component_index, (eigenvalue, vector) in enumerate(eigenpairs):
        singular_value = math.sqrt(eigenvalue * denominator)
        for sample, coordinate in zip(sample_names, vector):
            samples[sample].append(coordinate * singular_value)
        loadings = [
            {"feature": feature, "loading": _dot(row, vector) / singular_value}
            for feature, row in zip(feature_ids, rows)
        ]
        loadings.sort(key=lambda item: (-item["loading"], item["feature"]))
        top_positive = [item for item in loadings if item["loading"] > 0.0][:10]
        top_negative = [item for item in reversed(loadings) if item["loading"] < 0.0][:10]
        components.append(
            {
                "component": component_index + 1,
                "eigenvalue": eigenvalue,
                "explained_variance_percent": min(
                    max(eigenvalue / total_variance * 100.0, 0.0), 100.0
                ),
                "top_positive_loadings": top_positive,
                "top_negative_loadings": top_negative,
            }
        )

    warnings: list[str] = []
    if constant_features != 0:
        warnings.append(
            f"{constant_features} constant features contributed no PCA variance"
        )
    if component_limit < components_requested:
        warnings.append(
            f"requested {components_requested} components but matrix rank permits "
            f"at most {component_limit}"
        )

    return {
        "feature_count": len(feature_ids),
        "sample_count": len(sample_names),
        "scaled_features": scale_features,
        "total_variance": total_variance,
        "components": components,
        "samples": [
            {"sample": sample, "scores": scores} for sample, scores in samples.items()
        ],
        "warnings": warnings,
    }

File: workflows/org.linxira.benchmark-r/src/implementations/expression_pca_v1.R
# expression.pca.v1 — independent base-R implementation (no extra packages).
#
# A line-for-line port of the Rust engine's PCA
# (engine/crates/linxira-bio-core/src/expression.rs): same matrix reader
# rules, centering and optional scaling, the same deterministic power
# iteration on the sample Gram matrix (sinusoidal start vector, Gram-Schmidt
# against earlier components, identical convergence thresholds), and the same
# sign convention — the element with the largest absolute value is made
# positive (Rust's `max_by` returns the LAST maximum on ties, so the scan
# keeps the last one). prcomp is deliberately not used: its SVD sign
# convention and rounding differ, and this backend exists for a
# field-for-field comparison against the engine.

IMPLEMENTATION <- list(
  capability = "expression.pca.v1",
  input_roles = c("matrix"),
  parameters = c("components", "scale_features"),
  packages = character(0),
  software = function() list(),
  run = function(inputs, parameters) {
    components_requested <- as.integer(parameters$components %||% 2)
    if (is.na(components_requested) || components_requested <= 0) {
      implementation_error("components must be at least 1")
    }
    scale_features <- isTRUE(parameters$scale_features)
    matrix_data <- read_pca_matrix(inputs[["matrix"]])
    if (length(matrix_data$samples) < 2L) {
      implementation_error("PCA requires at least two samples")
    }
    centered <- centered_rows(matrix_data$values, scale_features)
    denominator <- length(matrix_data$samples) - 1
    total_variance <- sum(vapply(
      centered$rows,
      function(row) sum(row * row),
      numeric(1)
    )) / denominator
    if (total_variance <= .Machine$double.eps) {
      implementation_error("PCA requires at least one feature with non-zero variance")
    }
    component_limit <- min(components_requested, length(matrix_data$samples) - 1L, length(matrix_data$features))
    eigenpairs <- leading_eigenpairs(centered$rows, component_limit, denominator)
    if (length(eigenpairs) == 0L) {
      implementation_error("PCA could not resolve a non-zero component")
    }

    sample_scores <- setNames(
      vector("list", length(matrix_data$samples)),
      matrix_data$samples
    )
    components <- vector("list", length(eigenpairs))
    for (index in seq_along(eigenpairs)) {
      eigenvalue <- eigenpairs[[index]]$eigenvalue
      vector <- eigenpairs[[index]]$vector
      singular_value <- sqrt(eigenvalue * denominator)
      for (position in seq_along(matrix_data$samples)) {
        sample_scores[[position]] <- c(
          sample_scores[[position]],
          vector[[position]] * singular_value
        )
      }
      loadings <- data.frame(
        feature = matrix_data$features,
        loading = vapply(matrix_data$rows, function(row) dot(row, vector) / singular_value, numeric(1)),
        stringsAsFactors = FALSE
      )
      loadings <- loadings[order(-loadings$loading, loadings$feature, method = "radix"), ]
      # The engine takes the first 10 positive loadings of the descending
      # list and the first 10 negative ones of the REVERSED list, so tie
      # order inside both tails must follow that exactly. `rev()` on a
      # data.frame would reverse the columns, hence the explicit row index.
      top_positive <- head(loadings[loadings$loading > 0, ], 10L)
      negatives <- loadings[loadings$loading < 0, ]
      top_negative <- if (nrow(negatives) == 0L) negatives else head(
        negatives[rev(seq_len(nrow(negatives))), ],
        10L
      )
      components[[index]] <- list(
        component = index,
        eigenvalue = eigenvalue,
        explained_variance_percent = min(max(eigenvalue / total_variance * 100, 0), 100),
        top_positive_loadings = named_pair_list(top_positive),
        top_negative_loadings = named_pair_list(top_negative)
      )
    }

    warnings_out <- list()
    if (centered$constant_features != 0L) {
      warnings_out <- c(warnings_out, paste0(
        centered$constant_features, " constant features contributed no PCA variance"
      ))
    }
    if (component_limit < components_requested) {
      warnings_out <- c(warnings_out, paste0(
        "requested ", components_requested,
        " components but matrix rank permits at most ", component_limit
      ))
    }

    list(
      feature_count = length(matrix_data$features),
      sample_count = length(matrix_data$samples),
      scaled_features = scale_features,
      total_variance = total_variance,
      components = components,
      samples = unname(Map(
        function(sample, scores) list(sample = sample, scores = as.list(scores)),
        matrix_data$samples,
        sample_scores
      )),
      warnings = warnings_out
    )
  }
)

`%||%` <- function(left, right) if (is.null(left)) right else left

named_pair_list <- function(frame) {
  # unname: the JSON side must see an ARRAY of {feature, loading} objects,
  # not an object keyed by feature id.
  unname(Map(
    function(feature, loading) list(feature = feature, loading = loading),
    frame$feature,
    frame$loading
  ))
}

count_char <- function(text, character) {
  matches <- gregexpr(character, text, fixed = TRUE)[[1L]]
  if (length(matches) == 1L && matches[[1L]] == -1L) 0L else length(matches)
}

read_pca_matrix <- function(path) {
  magic <- readBin(path, "raw", n = 2L)
  connection <- if (length(magic) == 2L && identical(magic, as.raw(c(0x1f, 0x8b)))) {
    gzfile(path, "rt")
  } else {
    file(path, "rt")
  }
  lines <- tryCatch(readLines(connection, warn = FALSE), finally = close(connection))
  nonblank <- lines[nzchar(trimws(lines))]
  if (length(nonblank) == 0L) implementation_error("expression matrix is empty")
  delimiter <- if (count_char(nonblank[[1L]], "\t") >= count_char(nonblank[[1L]], ",")) "\t" else ","
  split_lines <- strsplit(nonblank, delimiter, fixed = TRUE)
  headers <- trimws(split_lines[[1L]])
  if (length(headers) < 2L || !nzchar(headers[[1L]])) {
    implementation_error("expected a named feature identifier column and at least one sample")
  }
  samples <- headers[-1L]
  if (any(!nzchar(samples)) || anyDuplicated(samples) != 0L) {
    implementation_error("sample names must be non-empty and unique")
  }
  features <- character(0)
  rows <- list()
  seen <- character(0)
  for (row_number in seq_along(split_lines)[-1L]) {
    record <- split_lines[[row_number]]
    if (length(record) != length(headers)) {
      implementation_error(paste0(
        "record ", row_number, " has ", length(record),
        " fields, expected ", length(headers)
      ))
    }
    feature <- trimws(record[[1L]])
    if (!nzchar(feature) || feature %in% seen) {
      implementation_error(paste0("feature identifier ", feature, " is empty or duplicated"))
    }
    seen <- c(seen, feature)
    row <- numeric(length(samples))
    for (position in seq_along(samples)) {
      value <- trimws(record[[position + 1L]])
      if (value == "" || value == "." || tolower(value) %in% c("na", "nan")) {
        implementation_error(paste0(
          "sample ", samples[[position]], " contains a missing value; ",
          "impute or filter it before this analysis"
        ))
      }
      parsed <- suppressWarnings(as.numeric(value))
      if (is.na(parsed)) {
        implementation_error(paste0(
          "sample ", samples[[position]], " contains non-numeric value ", value
        ))
      }
      if (!is.finite(parsed)) {
        implementation_error(paste0("sample ", samples[[position]], " contains a non-finite value"))
      }
      row[[position]] <- parsed
    }
    features <- c(features, feature)
    rows <- c(rows, list(row))
  }
  if (length(rows) == 0L) {
    implementation_error("expression matrix contains no feature rows")
  }
  list(features = features, samples = samples, rows = rows, values = rows)
}

centered_rows <- function(values, scale) {
  constant_features <- 0L
  rows <- vector("list", length(values))
  for (index in seq_along(values)) {
    row <- values[[index]]
    mean <- sum(row) / length(row)
    centered <- row - mean
    sum_squares <- sum(centered * centered)
    if (sum_squares <= .Machine$double.eps) {
      constant_features <- constant_features + 1L
      centered[] <- 0
    } else if (isTRUE(scale)) {
      denominator <- max(length(row) - 1L, 1L)
      centered <- centered / sqrt(sum_squares / denominator)
    }
    rows[[index]] <- centered
  }
  list(rows = rows, constant_features = constant_features)
}

dot <- function(left, right) sum(left * right)

orthogonalize <- function(vector, basis) {
  for (direction in basis) {
    projection <- dot(vector, direction)
    vector <- vector - projection * direction
  }
  vector
}

normalize_vector <- function(vector) {
  norm <- sqrt(dot(vector, vector))
  if (norm > .Machine$double.eps) vector <- vector / norm
  list(vector = vector, norm = norm)
}

gram_multiply <- function(rows, vector, denominator) {
  result <- numeric(length(vector))
  for (row in rows) {
    projection <- dot(row, vector) / denominator
    result <- result + row * projection
  }
  result
}

leading_eigenpairs <- function(rows, component_count, denominator) {
  if (component_count <= 0L) return(list())
  sample_count <- length(rows[[1L]])
  eigenvectors <- list()
  eigenpairs <- list()
  for (component in seq_len(component_count) - 1L) {
    indices <- seq_len(sample_count)
    phase <- indices * (component + 2L)
    vector <- sin(phase) + cos(phase * 0.37)
    vector <- orthogonalize(vector, eigenvectors)
    normalized <- normalize_vector(vector)
    vector <- normalized$vector
    if (normalized$norm <= .Machine$double.eps) {
      found <- FALSE
      for (basis in indices) {
        vector <- rep(0, sample_count)
        vector[[basis]] <- 1
        vector <- orthogonalize(vector, eigenvectors)
        normalized <- normalize_vector(vector)
        vector <- normalized$vector
        if (normalized$norm > .Machine$double.eps) {
          found <- TRUE
          break
        }
      }
      if (!found) break
    }
    for (iteration in 1:500) {
      next_vector <- gram_multiply(rows, vector, denominator)
      next_vector <- orthogonalize(next_vector, eigenvectors)
      normalized <- normalize_vector(next_vector)
      next_vector <- normalized$vector
      if (normalized$norm <= 1e-14) break
      alignment <- abs(dot(vector, next_vector))
      vector <- next_vector
      if (1.0 - alignment < 1e-11) break
    }
    projected <- gram_multiply(rows, vector, denominator)
    eigenvalue <- dot(vector, projected)
    if (!is.finite(eigenvalue) || eigenvalue <= 1e-12) break
    largest_positions <- which(abs(vector) == max(abs(vector)))
    largest <- largest_positions[[length(largest_positions)]]
    if (vector[[largest]] < 0) vector <- -vector
    eigenvectors <- c(eigenvectors, list(vector))
    eigenpairs <- c(eigenpairs, list(list(eigenvalue = eigenvalue, vector = vector)))
  }
  eigenpairs
}
